File: packages/typescript/src/fixtures.ts
import { readdirSync, readFileSync, statSync } from "node:fs";
import { join } from "node:path";

import type { Grid } from "./types.ts";

export type FixtureMeta = {
  id: string;
  title: string;
  description: string;
};

export type Fixture = {
  id: string;
  dir: string;
  meta: FixtureMeta;
  input: Grid;
};

/**
 * Absolute path to the language-neutral corpus root, resolved relative to this
 * source file so it works regardless of process cwd.
 */
export function corpusRoot(): string {
  // src/fixtures.ts → packages/typescript/src → ../.. → packages/typescript
  // → ../.. → repo root → fixtures/corpus
  return join(import.meta.dirname, "..", "..", "..", "fixtures", "corpus");
}

export function loadFixtures(root: string = corpusRoot()): Fixture[] {
  const ids = readdirSync(root)
    .filter((name) => {
      const dir = join(root, name);
      return statSync(dir).isDirectory();
    })
    .sort();
  return ids.map((id) => {
    const dir = join(root, id);
    const input = JSON.parse(
      readFileSync(join(dir, "input.json"), "utf8"),
    ) as Grid;
    const meta = JSON.parse(
      readFileSync(join(dir, "meta.json"), "utf8"),
    ) as FixtureMeta;
    return { id, dir, meta, input };
  });
}

export const goldenFiles = {
  anchorString: "anchor.string.txt",
  anchorJson: "anchor.json",
  anchorTokens: "anchor.tokenEstimate.txt",
  invertedIndexString: "invertedIndex.string.txt",
  invertedIndexJson: "invertedIndex.json",
  invertedIndexTokens: "invertedIndex.tokenEstimate.txt",
  rawBaselineTokens: "rawBaseline.tokenEstimate.txt",
} as const;

File: packages/python/sheet_compressor/fixtures.py
"""Helpers for loading the shared `fixtures/corpus/` corpus."""

import json
import os
from typing import List


GOLDEN_FILES = {
    "anchor_string": "anchor.string.txt",
    "anchor_json": "anchor.json",
    "anchor_tokens": "anchor.tokenEstimate.txt",
    "inverted_index_string": "invertedIndex.string.txt",
    "inverted_index_json": "invertedIndex.json",
    "inverted_index_tokens": "invertedIndex.tokenEstimate.txt",
    "format_aggregation_string": "formatAggregation.string.txt",
    "format_aggregation_json": "formatAggregation.json",
    "format_aggregation_tokens": "formatAggregation.tokenEstimate.txt",
    "charts": "charts.json",
    "raw_baseline_tokens": "rawBaseline.tokenEstimate.txt",
}


def corpus_root() -> str:
    """Resolve packages/python/sheet_compressor/fixtures.py -> repo root /fixtures/corpus."""
    here = os.path.dirname(os.path.abspath(__file__))
    return os.path.abspath(os.path.join(here, "..", "..", "..", "fixtures", "corpus"))


def load_fixtures(root: str = "") -> List[dict]:
    root = root or corpus_root()
    ids = sorted(
        name
        for name in os.listdir(root)
        if os.path.isdir(os.path.join(root, name))
    )
    out = []
    for fixture_id in ids:
        directory = os.path.join(root, fixture_id)
        with open(os.path.join(directory, "input.json"), "r", encoding="utf-8") as f:
            input_grid = json.load(f)
        with open(os.path.join(directory, "meta.json"), "r", encoding="utf-8") as f:
            meta = json.load(f)
        out.append({"id": fixture_id, "dir": directory, "meta": meta, "input": input_grid})
    return out
